samza-core/src/main/java/org/apache/samza/execution/JobNode.java

```diff
@@ -19,361 +19,150 @@
-import org.apache.samza.config.TaskConfigJava;
-import org.apache.samza.operators.OperatorSpecGraph;
 import org.apache.samza.operators.spec.InputOperatorSpec;
-import org.apache.samza.operators.spec.JoinOperatorSpec;
 import org.apache.samza.operators.spec.OperatorSpec;
-import org.apache.samza.operators.spec.OutputStreamImpl;
-import org.apache.samza.operators.spec.StatefulOperatorSpec;
-import org.apache.samza.operators.spec.WindowOperatorSpec;
-import org.apache.samza.table.TableConfigGenerator;
-import org.apache.samza.util.MathUtil;
 import org.apache.samza.serializers.Serde;
-import org.apache.samza.serializers.SerializableSerde;
 import org.apache.samza.table.TableSpec;
-import org.apache.samza.util.StreamUtil;
-import org.apache.samza.util.Util;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.google.common.base.Joiner;
-
-
 /**
  * A JobNode is a physical execution unit. In RemoteExecutionEnvironment, it's a job that will be submitted
  * to remote cluster. In LocalExecutionEnvironment, it's a set of StreamProcessors for local execution.
  * A JobNode contains the input/output, and the configs for physical execution.
  */
 public class JobNode {
   private static final Logger log = LoggerFactory.getLogger(JobNode.class);
-  private static final String CONFIG_INTERNAL_EXECUTION_PLAN = "samza.internal.execution.plan";
 
   private final String jobName;
   private final String jobId;
   private final String id;
-  private final OperatorSpecGraph specGraph;
-  private final List<StreamEdge> inEdges = new ArrayList<>();
-  private final List<StreamEdge> outEdges = new ArrayList<>();
-  private final List<TableSpec> tables = new ArrayList<>();
   private final Config config;
-
-  JobNode(String jobName, String jobId, OperatorSpecGraph specGraph, Config config) {
+  private final JobNodeConfigureGenerator configGenerator;
+  private final Map<String, StreamEdge> inEdges = new HashMap<>();
+  private final Map<String, StreamEdge> outEdges = new HashMap<>();
+  private final Map<String, TableSpec> tables = new HashMap<>();
+  private final ApplicationDescriptorImpl<? extends ApplicationDescriptor> appDesc;
+
+  JobNode(String jobName, String jobId, Config config, ApplicationDescriptorImpl appDesc,
+      JobNodeConfigureGenerator configureGenerator) {
     this.jobName = jobName;
     this.jobId = jobId;
     this.id = createId(jobName, jobId);
-    this.specGraph = specGraph;
     this.config = config;
+    this.appDesc = appDesc;
+    this.configGenerator = configureGenerator;
   }
 
-  public static Config mergeJobConfig(Config fullConfig, Config generatedConfig) {
-    return new JobConfig(Util.rewriteConfig(extractScopedConfig(
-        fullConfig, generatedConfig, String.format(JobConfig.CONFIG_JOB_PREFIX(), new JobConfig(fullConfig).getName().get()))));
-  }
-
-  public OperatorSpecGraph getSpecGraph() {
-    return this.specGraph;
+  static String createId(String jobName, String jobId) {
```

Comment:
Minor: Would prefer a more explicit method name reflecting what this ID is, since there are so many Ids.

Reply:
Make sense. I will change this to createJobNameAndId() together with getId() -> getJobNameAndId

Reply:
Another suggestion: Maybe jobUID since name + id uniquely identifies a job.